sdk/spring/spring-cloud-azure-autoconfigure/src/main/java/com/azure/spring/cloud/autoconfigure/aad/implementation/webapp/AadOAuth2UserService.java

```diff
@@ -83,46 +75,77 @@ public AadOAuth2UserService(AadAuthenticationProperties properties, RestTemplate
-                                  .map(AadAuthenticationProperties.UserGroupProperties::getAllowedGroupIds)
-                                  .orElseGet(Collections::emptySet);
-        DefaultOAuth2UserService oAuth2UserService = new DefaultOAuth2UserService();
-        oAuth2UserService.setRestOperations(createOAuth2ErrorResponseHandledRestTemplate(restTemplateBuilder));
-        this.oidcUserService = new OidcUserService();
-        this.oidcUserService.setOauth2UserService(oAuth2UserService);
+            .map(AadAuthenticationProperties::getUserGroup)
+            .map(AadAuthenticationProperties.UserGroupProperties::getAllowedGroupIds)
+            .orElseGet(Collections::emptySet);
         this.graphClient = graphClient;
     }
 
     /**
-     * Returns an {@link OAuth2User} after obtaining the user attributes of the End-User
-     * from the UserInfo Endpoint.
+     * Returns a {@link DefaultOidcUser} instance.
+     * <p/>
+     *
+     * The {@link DefaultOidcUser} instance is constructed with {@link GrantedAuthority}, {@link OidcIdToken} and nameAttributeKey.
+     * <a href="https://learn.microsoft.com/azure/active-directory/develop/userinfo#consider-using-an-id-token-instead">Azure AD</a> suggests get userinfo from idToken instead from the UserInfo Endpoint,
+     * this implementation will not get userinfo from the UserInfo Endpoint. Calling {@link org.springframework.security.oauth2.core.oidc.user.OidcUser#getUserInfo()} with the return instance will return null.
+     *
+     * <p/>
      *
      * @param userRequest the user request
-     * @return an {@link OAuth2User}
-     * @throws OAuth2AuthenticationException if an error occurs while attempting to obtain
-     * the user attributes from the UserInfo Endpoint
+     *
+     * @return a {@link DefaultOidcUser} instance.
+     *
+     * @throws OAuth2AuthenticationException if an error occurs.
      */
     @Override
     public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2AuthenticationException {
-        // Delegate to the default implementation for loading a user
-        OidcUser oidcUser = oidcUserService.loadUser(userRequest);
-        OidcIdToken idToken = oidcUser.getIdToken();
-        Set<String> authorityStrings = new HashSet<>();
-        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
+        Assert.notNull(userRequest, "userRequest cannot be null");
+
         ServletRequestAttributes attr = (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
         HttpSession session = attr.getRequest().getSession(true);
+        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
 
         if (authentication != null) {
             LOGGER.debug("User {}'s authorities saved from session: {}.", authentication.getName(), authentication.getAuthorities());
             return (DefaultOidcUser) session.getAttribute(DEFAULT_OIDC_USER);
         }
```

Comment:
Why not move this checking up to the top?

Reply:
can you make it more clear?

Reply:
I mean the checking for `authentication != null` can be done early, like the below code:

```java
public OidcUser loadUser(OidcUserRequest userRequest) throws OAuth2AuthenticationException {
        Authentication authentication = SecurityContextHolder.getContext().getAuthentication();
        if (authentication != null) {
            LOGGER.debug("User {}'s authorities saved from session: {}.", authentication.getName(), authentication.getAuthorities());
            return (DefaultOidcUser) session.getAttribute(DEFAULT_OIDC_USER);
        }

        ServletRequestAttributes attr = (ServletRequestAttributes) RequestContextHolder.currentRequestAttributes();
        HttpSession session = attr.getRequest().getSession(true);
        // ...
        return defaultOidcUser;
    }
```

Reply:
<img width="861" alt="image" src="https://user-images.githubusercontent.com/4465723/207798822-1320b052-d9dc-43dd-990c-a09e4bc3bb80.png">

We should get the session object first.